File: LeetCode/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#
# https://leetcode.com/problems/longest-palindromic-substring/description/
#
# algorithms
# Medium (35.50%)
# Likes:    30589
# Dislikes: 1887
# Total Accepted:    3.8M
# Total Submissions: 10.6M
# Testcase Example:  '"babad"'
#
# Given a string s, return the longest palindromic substring in s.
# 
# 
# Example 1:
# 
# 
# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.
# 
# 
# Example 2:
# 
# 
# Input: s = "cbbd"
# Output: "bb"
# 
# 
# 
# Constraints:
# 
# 
# 1 <= s.length <= 1000
# s consist of only digits and English letters.
# 
# 
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:  # 如果字符串长度为 1 或更短，直接返回
            return s

        start, max_length = 0, 0  # 记录最长回文子串的起始位置和长度

        # 定义一个内部函数，用于从中心扩展
        def expand_around_center(left: int, right: int):
            nonlocal start, max_length  # 使用外部变量
            # 当左右指针在字符串范围内，且字符相等时，继续扩展
            while left >= 0 and right < len(s) and s[left] == s[right]:
                current_length = right - left + 1  # 当前回文子串的长度
                if current_length > max_length:  # 如果当前回文更长，更新结果
                    start = left
                    max_length = current_length
                left -= 1  # 左指针左移
                right += 1  # 右指针右移

        # 遍历字符串中的每个字符
        for i in range(len(s)):
            # 奇数长度回文（中心是一个字符）
            expand_around_center(i, i)
            # 偶数长度回文（中心是两个字符之间）
            expand_around_center(i, i + 1)

        # 返回最长回文子串
        return s[start:start + max_length]
# @lc code=end

File: 00LeetCodeCpp/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
public:
    int expandAroundCenter(const string& s, int left, int right) {
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        return right - left - 1;

    }
    string longestPalindrome(string s) {
        // if (s.length() < 2) return s;
        // int start = 0, maxLen = 0;

        // for (int idx = 0; idx < s.length(); idx++) {
        //     if (s.length() - idx <= maxLen / 2) break;
        //     int curLen = max(expandAroundCenter(s, idx, idx),
        //                      expandAroundCenter(s, idx, idx + 1));

        //     if (curLen > maxLen) {
        //         maxLen = curLen;
        //         start = idx - (curLen - 1) / 2;
        //     }
        // }

        /* */
        // int n = s.length();
        // if (n < 2) return s;
        
        // int start = 0, maxLen = 1;
        // for (int i = 0; i < n; ) {
        //     // 剪枝：剩余字符不够长了，直接收工
        //     if (n - i <= maxLen / 2) break;

        //     int left = i, right = i;
        //     // 技巧：跳过中间所有重复字符（如 "baaaaab" 中的 "aaaaa"）
        //     while (right < n - 1 && s[right + 1] == s[right]) {
        //         right++;
        //     }
            
        //     // 下一次循环直接从重复字符后面开始，大大减少循环次数
        //     i = right + 1;

        //     // 从这个“中心块”向两边扩散
        //     while (left > 0 && right < n - 1 && s[left - 1] == s[right + 1]) {
        //         left--;
        //         right++;
        //     }

        //     int curLen = right - left + 1;
        //     if (curLen > maxLen) {
        //         start = left;
        //         maxLen = curLen;
        //     }
        // }

        /* Manacher */
        // 1. 预处理：插入 #
        string t = "$#";
        for (char c : s) { t += c; t += '#'; }
        t += '@'; // 首尾不同字符防止越界

        int n = t.size();
        vector<int> p(n, 0);
        int C = 0, R = 0, maxLen = 0, centerIndex = 0;

        for (int i = 1; i < n - 1; i++) {
            // 【核心代码】利用对称性初始化 P[i]
            if (i < R) p[i] = min(R - i, p[2 * C - i]);
            
            // 尝试继续向外扩散
            while (t[i + p[i] + 1] == t[i - p[i] - 1]) p[i]++;

            // 更新中心和最右边界
            if (i + p[i] > R) {
                C = i;
                R = i + p[i];
            }

            if (p[i] > maxLen) {
                maxLen = p[i];
                centerIndex = i;
            }
        }
        // 计算原字符串起始位置并返回
        int start = (centerIndex - maxLen) / 2;
        return s.substr(start, maxLen);
    }
};
// @lc code=end

